src/Filesystem/Filesystem.php

```diff
@@ -0,0 +1,194 @@
+            | FilesystemIterator::SKIP_DOTS;
+        $directory = new FilesystemIterator($path, $flags);
+
+        if ($filter === null) {
+            return $directory;
+        }
+
+        if (is_string($filter)) {
+            return new RegexIterator($directory, $filter);
+        }
+
+        return new CallbackFilterIterator($directory, $filter);
+    }
+
+    public function findRecursive(string $path, $filter = null, ?int $flags = null): Traversable
+    {
+        $flags = $flags ?? FilesystemIterator::KEY_AS_PATHNAME
+            | FilesystemIterator::CURRENT_AS_FILEINFO
+            | FilesystemIterator::SKIP_DOTS;
+        $directory = new RecursiveDirectoryIterator($path, $flags);
+
+        $dirFilter = new RecursiveCallbackFilterIterator(
+            $directory,
+            function (SplFileInfo $current) {
+                if ($current->getFilename()[0] === '.' && $current->isDir()) {
+                    return false;
+                }
+
+                return true;
+            }
+        );
+
+        $flatten = new RecursiveIteratorIterator(
+            $dirFilter,
+            RecursiveIteratorIterator::CHILD_FIRST
+        );
+
+        if ($filter === null) {
+            return $flatten;
+        }
+
+        if (is_string($filter)) {
+            return new RegexIterator($flatten, $filter);
+        }
+
+        return new CallbackFilterIterator($flatten, $filter);
+    }
+
+    public function dumpFile(string $filename, string $content): void
+    {
+        $dir = dirname($filename);
+        if (!is_dir($dir)) {
+            $this->mkdir($dir);
+        }
+
+        $exits = file_exists($filename);
+
+        if ($this->isStream($filename)) {
+            file_put_contents($filename, $content);
+        } else {
+            file_put_contents($filename, $content, LOCK_EX);
+        }
+
+        if (!$exits) {
+            chmod($filename, 0666 & ~umask());
+        }
+    }
+
+    public function mkdir(string $dir, int $mode = 0755): void
+    {
+        if (is_dir($dir)) {
+            return;
+        }
+
+        $old = umask(0);
+        // @codingStandardsIgnoreLine
+        if (@mkdir($dir, $mode, true) === false) {
+            umask($old);
+            throw new Exception(sprintf('Failed to create directory "%s"', $dir));
+        }
+
+        umask($old);
+    }
+
+    public function deleteDir(string $path): bool
+    {
+        if (!file_exists($path)) {
+            return true;
+        }
+
+        if (!is_dir($path)) {
+            throw new Exception(sprintf('"%s" is not a directory', $path));
+        }
+
+        $iterator = new RecursiveIteratorIterator(
+            new RecursiveDirectoryIterator($path, FilesystemIterator::SKIP_DOTS),
+            RecursiveIteratorIterator::CHILD_FIRST
+        );
+
+        $result = true;
+        foreach ($iterator as $fileInfo) {
+            switch ($fileInfo->getType()) {
+                case 'dir':
```

Comment:
Does PHP give us constants for these strings?

Reply:
Not that I know of, will check.

Reply:
Add `RecursiveDirectoryIterator::CURRENT_AS_SELF` flag to `RecursiveDirectoryIterator` constructor. 

`$fileInfo` is becomes instance of `RecursiveDirectoryIterator`, and has 

-    `DirectoryIterator::isDir()` - Determine if current DirectoryIterator item is a directory
-    `DirectoryIterator::isDot()` - Determine if current DirectoryIterator item is '.' or '..'
-    `DirectoryIterator::isFile()` - Determine if current DirectoryIterator item is a regular file
-    `DirectoryIterator::isLink()` - Determine if current DirectoryIterator item is a symbolic link